src/task4feedback/simulator/utility.py

```diff
@@ -115,6 +111,220 @@ def get_heft_rank(task):
+            task_data_list = (task.info.data_dependencies[AccessType.READ] +
+                              task.info.data_dependencies[AccessType.WRITE] +
+                              task.info.data_dependencies[AccessType.READ_WRITE])
+            for data in task_data_list:
+                data_info = scheduler_state.objects.datamap[data.id].info
+                initial_placement = data_info.location
+                if initial_placement not in data_cache:
+                    data_cache[initial_placement] = set()
+                data_cache[initial_placement].add(data_info.id) 
+
+    print("data cache:", data_cache)
+
+    HEFTEvent = namedtuple("HEFTEvent", "task start end")
+
+    def get_start_time(heft_event):
+        return heft_event.start
+
+    max_heft = -1
+    if update_task_order:
+        tasklist[:] = []
+        heft_events = []
+    # Forward phase to allocate each task to each device
+    for task in heft_sorted_tasks:
+        duration = convert_to_float(
+            scheduler_state.get_task_duration(
+                task, task.info.runtime.locations[0]
+            ).scale_to("ms")
+        )
+
+        # Get task's data dependencies
+        if task.info.data_dependencies is not None:
+            read = task.info.data_dependencies[AccessType.READ]
+            write = task.info.data_dependencies[AccessType.WRITE]
+            rw = task.info.data_dependencies[AccessType.READ_WRITE]
+            src_data = read + write + rw
+
+        ready_time = 0
+        earliest_start = None
+        earliest_start_agent = -1
+
+        # Try to insert each task to each agent (device)
+        for agent_id, agent in agents.items():
+
+            # Collect unique data to be moved from dependency tasks
+            # This is used to identify data from initial placement
+            moved_data_from_dependencies = set()
+
+            current_device = Device(Architecture.GPU, agent_id)
+
+            # Iterate all dependencies and check rank + communication overhead
+            # and find task's ready time when the task is assigned to the current
+            # agent
+            for dep in task.dependencies:
+                dependency_instance = taskmap[dep]
+
+                if isinstance(dependency_instance, SimulatedDataTask):
+                    continue
+
+                # TODO(hc): We can reuse this information collected in the above later
+                # Compute communication overhead
+                all_dep_data = dependency_instance.info.data_dependencies.all_ids()
+                intersected_data = []
+                intersected_data_size = 0
+                for sd in src_data:  # SimulatedData
+                    # If the data is already planned to be materialized on this device,
+                    # skip it.
+                    if sd.id in data_cache[current_device]:
+                        continue
+
+                    for dd in all_dep_data:  # DataID
+                        if sd.id == dd:
```

Comment:
I think this loops still over counts data that is shared on multiple incoming edges. 
For example, a task that writes to block A that depends on two other tasks that both read block A

Reply:
I am aware of it but this consideration would require more policeis in HEFT. 
For example, which dependency should we consider?
We should consider which dependency would be the best option for this task. 
One dependency might be finished earlier than others. But due to its placement, data transfer time would be more expensive. We need to consider such factors and I think this would make over complex the policy. So for now, I just skipped it.